add ARIB caption decoder using libaribcaption

aimoff opened the pull request:
PR for review of patchset sent to FFmpeg ML.

---

This patch set add another ARIB caption decoder using libaribcaption
external library: https://github.com/xqq/libaribcaption
The library decodes subtitles of ISDB-based TV broadcasting.
It is not only for Japanese ARIB STD-B24 caption, but also for
Brazilian ABNT NBR 15606-1 and Philippines version of ISDB-T.

Unlike libaribb24, it supports 3 types of subtitle outputs:
* text: plain text
* ass: ASS formatted text
* bitmap: bitmap image

Default subtitle type is ASS as same as libaribb24, but it can
be changed by `-sub_type` option.
You can see ARIB caption with ffplay tool:
  ffplay -sub_type bitmap MPEG.TS

Sample files exist under:
  https://streams.videolan.org/streams/ts/ARIB/
Some of them are encrypted and some don't contain ARIB caption data.
Good samples for ARIB caption are:
  brazil/07-25_20-33-35_UCV - HD_.ts
  japan/channel2[137]_clear.ts
  japan/osaka_15.ts

aimoff committed:
lavc/libaribcaption.c: add ARIB caption decoder using libaribcaption

This patch add another ARIB caption decoder using libaribcaption
external library.

Unlike libaribb24, it supports 3 types of subtitle outputs:
* text: plain text
* ass: ASS formatted text
* bitmap: bitmap image

Default subtitle type is ass as same as libaribb24.

To compile with this feature:
* libaribcaption external library has to be pre-installed.
  https://github.com/xqq/libaribcaption
* configure with `--enable-libaribcaption` option.

`--enable-libaribb24` and `--enable-libaribcaption` options are
not exclusive. If both enabled, libaribcaption precedes as
order listed in `libavcodec/allcodecs.c`.

aimoff committed:
lavc/codec_desc.c: remove AV_CODEC_PROP_TEXT_SUB property

To support bitmap type of subtitles, remove AV_CODEC_PROP_TEXT_SUB
property from codec descriptor for AV_CODEC_ID_ARIB_CAPTION.
It is similar way to `libavcodec/libzvbi-teletextdec.c`
(AV_CODEC_ID_DVB_TELETEXT).

Instead, each subtitle decoder has to specify subtitile format.
`libavcodec/libaribb24.c` uses same AV_CODEC_ID_ARIB_CAPTION and
expects AV_CODEC_PROP_TEXT_SUB is defined. Thus add a line to
specify text format subtitle.

aimoff committed:
lavf/mpegts.c: set some properties for ARIB caption

Some additional properties are set for ARIB caption.
* need_parsing = 0
  ARIB caption doesn't require any parser.
  This avoids "parser not found" warning message.
* need_context_update = 1
  When any profiles are changed, set this flag to notify.
* codecpar->width / codecpar->height
  Find best video stream and set frame size for ARIB_PROFILE_A
  type of ARIB caption.

aimoff committed:
doc/decoders.texi: add document of aribcaption decoder

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+    return 0;
+}
+
+#if !defined(DEFAULT_SUBTITLE_TYPE)
+# define DEFAULT_SUBTITLE_TYPE SUBTITLE_ASS
+#endif
+#if !defined(ASS_WORKAROUND)
+# define ASS_WORKAROUND 1
+#endif
```
These should simply be set directly in the `AVOption` array.

aimoff replied:
These provide only default behavior.
Former one can be embeded in the `AVOption` array. The default behavior would not be changed.
Latter one depends when most of softwares (or major softwares) can handle multi-rectangle ASS subtitle. Thus, I prefer to put it out from the `AVOption` array.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+            const char *fonts = ctx->font;
+
+            while (!is_nomem && *fonts) {
+                const char **ff = av_realloc_array(font_families, count + 1, sizeof(*font_families));
+                if (!ff)
+                    is_nomem = 1;
+                else {
+                    font_families = ff;
+                    ff[count++] = av_get_token(&fonts, ",");
+                    if (!ff[count - 1])
+                        is_nomem = 1;
+                    else if (*fonts)
+                        fonts++;
+                }
```
If you get an allocation failure, you should return AVERROR(ENOMEM), rather than attempting to continue in a degraded state.

aimoff replied:
This does't continue in a degraded state, but return `AVERROR(ENOMEM)` after clean up.

```c
            while (count)
                av_freep(&font_families[--count]);
            av_freep(&font_families);
            if (is_nomem) {
                aribcaption_close(avctx);
                return AVERROR(ENOMEM);
            }
```

If such clean up is not needed (i.e., automatically freed), I can simply return with `AVERROR(ENOMEM)`.

rcombs replied:
You can add a label (e.g. `fail:`) and `goto` it on failure.

aimoff replied:
It is a matter of preference if it is not specified in coding rule.
I prefer to keep current way because same clean up routine runs regardless of error or not.

aimoff replied:
Instead of `while (!is_nomem && *fonts)` loop, I can change to add `break;` at allocation failure to quit from loop.

rcombs replied:
Sure, that's a bit cleaner.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+    if (ctx->force_stroke_text)
+        outline = (int)(ctx->stroke_width * 4.0 / 3.0);
+
+    if (fonts && *fonts)
+        font_name = av_get_token(&fonts, ",");
+    else
+        font_name = av_strdup(DEFAULT_FONT_FAMILY);
```
Why not just make `sans-serif` the default for the AVOption?

aimoff replied:
The `libaribcaption` library has it's default font for bitmap depending on operating system. I would like to follow it if `font` option is not specified.
On the other hand, we have to set default font for ASS subtitle.

rcombs replied:
This isn't using a default from libaribcaption, though? It's just using `DEFAULT_FONT_FAMILY`, which is defined as `sans-serif`.

aimoff replied:
`libaribcaption` renders (with default font) when `sub_type` is `bitmap` only.
When `sub_type` is `ass`, `libaribcaption` parses ARIB caption, but does not render nor encode to ASS text.
The `DEFAULT_FONT_FAMILY` is set to ASS header only, but most players such as `mpv` set its own ASS header and specified font is ignored.

rcombs replied:
Ahh, maybe name the macro `ASS_DEFAULT_FONT` or the like for clarity?

aimoff replied:
`ASS_DEFAULT_FONT` is already defined in `libavcodec/ass.h` but is `"Arial"`.
The font is not suitable for ARIB caption.
Which is your preference, redifine with same macro name or define another new (but ASS related) macro name?

rcombs replied:
Hmm, maybe `DEFAULT_FONT_ASS`? And add a comment indicating that it's only used for ASS, because the bitmap renderer has its own internal default.

rcombs commented on libavformat/mpegts.c:
```diff
@@ -2175,8 +2181,19 @@ int ff_parse_mpeg2_descriptor(AVFormatContext *fc, AVStream *st, int stream_type
+                st->codecpar->profile = picked_profile;
+                sti->need_context_update = 1;
+            }
+            if (v_st &&
+                (st->codecpar->width != v_st->codecpar->width ||
+                 st->codecpar->height != v_st->codecpar->height)) {
+                st->codecpar->width = v_st->codecpar->width;
+                st->codecpar->height = v_st->codecpar->height;
+                sti->need_context_update = 1;
+            }
```
This section isn't appropriate. Instead, provide an AVOption with type `AV_OPT_TYPE_IMAGE_SIZE` in your decoder, defaulting to the layout dimensions derived from the subtitle stream itself, and allow player apps to set it to an appropriate value (this is usually the _output_ resolution) when using the bitmap renderer.

aimoff replied:
I'll consider it.

aimoff replied:
I add `-canvas_size` option to specify input video size.
The same option already exists in `ffmpeg` tool, and it set `width` and `height` in `AVCodecContext`. Since `ffmpeg` tool doesn't pass `-canvas_size` option to `libaribcaption`, thus I use the values in `AVCodecContext`. The `-canvas_size` option in `AVOption` is used by other tools such as ffplay.

Note that the `canvas_size` specifies frame size of input video (i.e., SAR basis). Actual canvas size is stretched to 16:9 to match DAR-derived video size.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+        ctx->charstyle |= ARIBCC_CHARSTYLE_STROKE;
+
+    encode_scheme = ARIBCC_ENCODING_SCHEME_AUTO;
+    if (avctx->sub_charenc_mode == FF_SUB_CHARENC_MODE_DO_NOTHING)
```
This is meant to control libavcodec's built-in encoding conversion infrastructure. If you want to let callers control the encoding libaribcaption uses, you should provide an `AVOption` defaulting to `ARIBCC_ENCODING_SCHEME_AUTO`.

aimoff replied:
Encoding depends on broadcast source (i.e., country).
Most of cases, it should be determined automatically.
It is last resort when it cannot be determined automatically.
If `sub_charenc_mode` is not suitable means, I'll add one in `AVOption`.

aimoff replied:
Tentatively, I add `-encoding_scheme` option locally and specify it as same value as `aribcc_encoding_scheme_t`:
* auto
* jis
* utf8
* latin

I concern the option name might be too generic and it may conflct with other codecs.

rcombs replied:
Nice! If you want to be more specific, you can use `arib_text_encoding` or the like, but I don't think it's a big deal either way.

aimoff replied:
I determined the option name as `-caption_encoding`.
This should not be confused.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+        if (ret != 0) {
+            av_log(avctx, AV_LOG_ERROR, "Failed to set ASS header: %s\n",
+                                        av_err2str(ret));
+            aribcaption_close(avctx);
```
You have a lot of calls to `aribcaption_close` on error paths in this function. You can instead set `FF_CODEC_CAP_INIT_CLEANUP` in `caps_internal`, which will result in libavcodec automatically calling your `close` routine if initialization fails.

aimoff replied:
Thanks. I'll try it.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+
+    sub->format = 1; /* text */
+    if (ctx->caption.region_count == 0) {
+        /* clear previous caption for indefinite duration  */
+        ff_ass_add_rect(sub, "{\\r}", ctx->readorder++, 0, NULL, NULL);
+        return 1;
```
The `{\r}` shouldn't be necessary here; you should be able to just use an empty string.

rcombs commented on doc/decoders.texi:
```diff
@@ -353,6 +353,105 @@ Enabled by default.
+
+If not specified, use internaly defined default font family.
+
+@item -ass_workaround @var{boolean}
+Since some players (e.g., @dfn{mpv}) can't handle multi-rectangle ASS
+subtitle properly, default behavior is all the texts are displayed in
+a single-rectangle at a time.
```
What behavior are you seeing in mpv with this set to false? If this isn't working properly, it should be fixed (either in the decoder or in mpv).

aimoff replied:
There is a `mpv` implementation that can handle multi-rectangle ASS subtitle:
https://github.com/0p1pp1/mpv

You can find typical case in https://streams.videolan.org/streams/ts/ARIB/japan/osaka_15.ts around 00:03:35.
![osaka_15.ts](https://user-images.githubusercontent.com/19228344/218261058-31b5b752-bae9-4ee3-afeb-49575bddd453.png)
2 captions are displayed on each person at a same time.
You can see same scene with `ffplay -sub_type bitmap osaka_15.ts` after this patchset is applied.

Current plain `mpv` cannot handle multi-rectangle ASS subtitle.
If plain `mpv` plays it with `-ass_workaround false`, first caption is immediately clear after second caption is displayed; i.e., only second caption can be seen.

rcombs replied:
![image](https://user-images.githubusercontent.com/819638/218274447-1ef67ca4-be7d-4bf1-a478-f7dc7a218aa3.png)
This works fine in current mpv (mpv-player/mpv@7990dd8f3f664e01c984a405dad7c8af320cabf1) for me?

aimoff replied:
Oh, my knowledge is little bit old.
I found mpv release/0.35 can handle multi-rectangle ASS subtitles.

I would like change default value of `ass_workaround` to 0, but stay default definition to be outside of `AVOption`.

rcombs replied:
I don't understand why you want the default to be a macro.

Also, I'd suggest naming the option `single_rect` to clarify its behavior.

aimoff replied:
The reason is explained in later description (but should amend the explanation to revert default value).

```texi
The default is @var{true}.

If a player can handle multi-rectangle ASS subtitle and you prefer
more suitable positioning, set this option to @var{false} or define
@env{ASS_WORKAROUND=0} to change default behavior at compilation.
```

This can easily change default behavior without change the code.

aimoff replied:
I will change the option name to `ass_single_rect`.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+
+    aribcaption_flush(avctx);
+
+    if (ctx->clut)
+        av_freep(&ctx->clut);
```
You don't need to `if` around this; `av_freep(&x)` where `x` is null is a no-op.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+    if (ARIBCC_COLOR_DIFF_RGB(new_color, old_color))
+        av_bprintf(buf, "{\\%dc&H%06x&}", color_num,
+                                          ARIBCC_COLOR_RGB(new_color));
+    if (ARIBCC_COLOR_DIFF_A(new_color, old_color) && ARIBCC_COLOR_A(new_color))
```
This should check if the alpha is _different_; if it changes to 0, we should set that.

aimoff replied:
Actual ISDB broadcasting doesn't allways set alpha value, and 0 is set in many cases.
Alpha value 0 means transparent, and as the result subtitle cannot be seen if we strictly obey it.

rcombs replied:
0 in ASS means fully opaque; _0xFF_ means fully transparent. I think you need to invert the value here.

I don't see any corresponding code in libaribcaption's renderer to ignore 0 values?

aimoff replied:
Is it true?
If so, above line 559 should not changed. Instead, ASS style override code number would be `0xFF - alpha`.

rcombs replied:
Correct, though I'm still confused about why this check is necessary when libaribcaption's renderer doesn't seem to have anything equivalent (unless I'm missing something?). In any case, it at least needs a comment explaining why.

aimoff replied:
I checked some TS files on my hand. I can't find text color with alpha 0. So, I might see apparition.
I'll remove the check.

rcombs reviewed (changes requested):
A few initial comments.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+    FF_CODEC_DECODE_SUB_CB(aribcaption_decode),
+    .flush     = aribcaption_flush,
+    .p.priv_class= &aribcaption_class,
+    .caps_internal = FF_CODEC_CAP_INIT_THREADSAFE,
```
This patch doesn't build as-is, because `FF_CODEC_CAP_INIT_THREADSAFE` is now the default, and the macro has been removed.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1141 @@
+            else
+                ff_ass_bprint_text_event(&buf, ch->u8str, strlen(ch->u8str), "", 0);
+        }
+        av_bprintf(&buf, "{\\r}%s",
+                         (i + 1 < ctx->caption.region_count) ? "\\N" : "");
```
This line only makes sense on the `ass_workaround` path, and the `{\\r}` is only necessary if inserting the newline, so this could be:
```suggestion
        if (ctx->ass_workaround && (i + 1 < ctx->caption.region_count))
            av_bprintf(&buf, "{\\r}\\N");
```

aimoff replied:
The `{\\r}` is diverted from former `libaribb24`.
It seems it is not required for multi-rect ASS, I will change the code as you suggested.

aimoff committed:
Reflect review results from rcombs

* rename `DEFAULT_FONT_FAMILY` macro to `DEFAULT_FONT_ASS`.
* rename `ass_workaround` option to `ass_single_rect`.
* remove unnecessary check before `av_freep()`.
* remove unnecessary `"{\r}"` output at end of each ASS rect.
* remove `FF_CODEC_CAP_INIT_THREADSAFE` and set `FF_CODEC_CAP_INIT_CLEANUP`
  flag in `.caps_internal`, then remove unnecessary `aribcaption_close()`.
* remove `DEFAULT_SUBTITLE_TYPE` macro and embed default `-sub_type` option
  in `AVOption` structure.
* add `-caption_encoding` to specify encoding of subtitle text.
* specify input video size by `-canvas_size` option.
* some cosmetic changes.

aimoff commented:
I pushed latest code that reflect @rcombs comments.

aimoff committed:
Some minor changes applyed

* change behavior of ASS subtitle with PROFILE_C to be single rectangle.
* document more detail of `-canvas_size` option.
* some cosmetic changes.

aimoff commented:
I changed behavior of ARIB caption with PROFILE_C.
Since FFmpeg anounced schedule of version 6.0, I have plan to send the patchset to ML in this weekend.
If you are maintainer of FFmpeg, please make review comment on ML and push the patchset.

rcombs commented on doc/decoders.texi:
```diff
@@ -386,25 +386,45 @@ Simple text based output without formatting.
+positions at a time (multi-rectangle subtitle).
+Since some players (e.g., old @dfn{mpv}) can't handle multi-rectangle ASS
+subtitle properly, this option can change the behavior so that all the texts
+are displayed in a single-rectangle.
```
"can't handle multiple ASS rectangles in a single AVSubtitle, or multiple ASS rectangles of indeterminate duration with the same start timestamp"

also, "in a single rectangle"

rcombs commented on doc/decoders.texi:
```diff
@@ -386,25 +386,45 @@ Simple text based output without formatting.
-If a player can handle multi-rectangle ASS subtitle and you prefer
-more suitable positioning, set this option to @var{false} or define
-@env{ASS_WORKAROUND=0} to change default behavior at compilation.
+If your player cannot handle multi-rectangle ASS subtitle properly,
```
"cannot handle AVSubtitles with multiple ASS rectangles properly,"

rcombs commented on doc/decoders.texi:
```diff
@@ -438,6 +458,9 @@ Specify whether to render replaced DRCS characters as Unicode characters.
 The default is @var{true}.
 
+@item -canvas_size @var{image_size}
+Specify input video size to determine size of canvas to render subtitle.
```
"Specify the resolution of the canvas to render subtitles to; usually, this should be either the resolution of the video or the resolution it's being displayed at. This only applies when @var{subtitle_type} is set to @var{bitmap}."

aimoff replied:
First sentence is still under considering as discussed below.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -67,7 +67,8 @@ typedef struct ARIBCaptionContext {
 
     enum AVSubtitleType subtitle_type;
-    bool ass_workaround;
+    enum aribcc_encoding_scheme_t encoding_scheme;
```
This should be an `int`, since `enum aribcc_encoding_scheme_t` isn't guaranteed to have the same size as `int` across all platforms.

rcombs commented on doc/decoders.texi:
```diff
@@ -461,6 +461,26 @@ The default is @var{true}.
 @item -canvas_size @var{image_size}
 Specify input video size to determine size of canvas to render subtitle.
 
+The aribcaption decoder assumes input frame size for bitmap rendering as bellow:
```
"below"

rcombs commented on doc/decoders.texi:
```diff
@@ -461,6 +461,26 @@ The default is @var{true}.
 Specify input video size to determine size of canvas to render subtitle.
 
+The aribcaption decoder assumes input frame size for bitmap rendering as bellow:
+@enumerate
+@item
+PROFILE_A : 1440 x 1080 with SAR 4:3
+@item
+PROFILE_C : 320 x 180 with SAR 1:1
+@end enumerate
+
+If actual frame aspect ratio (FAR) of input video does not match above assumption,
+the rendered captions may be distorted.
+To make the captions undistorted, add @code{-canvas_size} option to specify
+actual input video size.
```
It's not clear what this means; "frame aspect ratio" isn't a common term in ffmpeg. Is this saying that the `canvas_size` needs to match the output resolution?

aimoff replied:
> Is this saying that the `canvas_size` needs to match the output resolution?

No.
On most common usages, `-canvas_size` should be set to actual **input** video size, **if** the input video size does not match to above assumption.
If input video size is 720 x 540 with SAR 4:3, i.e. half size of assumption, the bitmap subtitle is automatically shrunk to fit actual video size without `-canvas_size` option.
But if input video size is 1920 x 1080 with SAR 1:1, rendered bitmap subtitle will be distorted unless `-canvas_size` is specified even if output video size (DAR: Display Aspect Ratio) is the same 1920 x 1080. In such cases, `-canvas_size` should be set to `1920x1080`. (A Japanese broadcaster uses this format.)

On few specific usages, `-canvas_size` can be set to larger than input video size in conjunction with `-filter_complex` to rescale only video output. For example, a 720 x 540 size video (SAR 4:3) can be rescaled with embeded bitmap subtitle as following 2 ways:
```
ffmpeg -sub_type bitmap -i INPUT.TS -filter_complex "[0:v][0:s]overlay[v]" -map "[v]" -s 1920x1080 OUTPUT.mp4
ffmpeg -sub_type bitmap -canvas_size 1920x1080 -i INPUT.TS -filter_complex "[0:v]scale=1920x1080,setsar=1[v], [v][0:s]overlay[v]" -map "[v]" OUTPUT.mp4
```
The latter one shows better sublitle quality.

aimoff replied:
My terminology is based on following "mpeg4-avc terminology":
https://forum.videohelp.com/threads/323530-please-explain-SAR-DAR-PAR

To translate above to "MPEG2 terminology", _SAR_ is to be PAR, and _FAR_ (_input video size_) is to be SAR. This may cause confusion.

aimoff replied:
I drop the words "frame aspect ratio".
Instead, "frame size" is used.

rcombs commented on doc/decoders.texi:
```diff
@@ -461,6 +461,26 @@ The default is @var{true}.
+To make the captions undistorted, add @code{-canvas_size} option to specify
+actual input video size.
+
+Note that the @code{-canvas_size} option is not required for video with
+different size but same aspect ratio.
```
Probably say "optional" instead of "not required", since it's still _useful_ to set for quality.

aimoff replied:
I can reword as your suggestion, but are there any more suitable brief explanations based on my previous comment?
I don't want to explain so details in the document.

rcombs commented on doc/decoders.texi:
```diff
@@ -461,6 +461,26 @@ The default is @var{true}.
+In such cases, the caption will be stretched or shrunk to actual video size
+if @code{-canvas_size} option is not specified.
+If @code{-canvas_size} option is specified with different size,
+the caption will be stretched or shrunk as specified size with calcurated SAR.
```
"calculated"

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -584,6 +584,8 @@ static int aribcaption_trans_ass_subtitle(ARIBCaptionContext *ctx)
         if ((ret = set_ass_header(ctx)) < 0)
             return ret;
     }
+    if (ctx->avctx->profile == FF_PROFILE_ARIB_PROFILE_C)
+        single_rect = true;
```
Why's this?

aimoff replied:
The profile C is defined for devices that have very limited resources.
Please refer [ARIB TR-B14](https://www.arib.or.jp/english/std_tr/broadcasting/tr-b14.html) version 3.8 Fascicle 1-(2/2) Volume 3 [Section 4] "Operational provisions related to C-profile". The section has been removed from version 6.7. The Table 3-1 explains that the caption may be up to 4 lines, and Table 3-2 explains the position is not provided by contents.
Thus, I think it is better to treat it as single ASS rectangle.

At present, mpv doesn't display ASS subtitle of profile C correctly even if this change is applied. I don't know the reason. Output of `ffmpeg -i INPUT.TS -map STREM_ID_OF_PROFILE_C OUTPUT.ass` seems to be correct, but mpv displays garbled characters.
In addition, similar change should be made for bitmap subtiles, but I don't have easy resolution.

rcombs replied:
Do you have a sample file I can take a look at?

If you decide to keep this, you should include a comment noting that profile C doesn't allow for positioning.

aimoff replied:
You can see it in the same TS file:
https://streams.videolan.org/streams/ts/ARIB/japan/osaka_15.ts
```
ffmpeg -fix_sub_duration -i osaka_15.ts -map 0:14 osaka_15.ass
mpv osaka_15.ts --vid=2 --aid=2 --sid=2
```
Recently @jeeb found the root cause and sent a patch to mpv:
https://github.com/mpv-player/mpv/pull/11340
You can see the subtitles by mpv with this patch.

rcombs replied:
Would definitely be worth reporting to @xqq. In any case, worth a comment for now :)

aimoff replied:
I already shared such information with @xqq.

With latest patch for this PR, you can see bitmap subtitle of profile C by ffplay.
```
ffplay -sub_type bitmap osaka_15.ts -vst 13 -ast 12 -sst 14
```

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1166 @@
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    if (!(avctx->flags2 & AV_CODEC_FLAG2_RO_FLUSH_NOOP)) {
```
This flag should only affect setting `readorder`; the rest of the function should run regardless.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1166 @@
+    aribcaption_flush(avctx);
+
+    av_freep(&ctx->clut);
+    if (ctx->renderer) {
+        aribcc_renderer_free(ctx->renderer);
+        ctx->renderer = NULL;
+    }
+    if (ctx->decoder) {
+        aribcc_decoder_free(ctx->decoder);
+        ctx->decoder = NULL;
+    }
+    if (ctx->context) {
+        aribcc_context_free(ctx->context);
+        ctx->context = NULL;
+     }
```
- none of these checks should be required; these functions are safe to pass NULL to
- it's not necessary to zero the pointers, since the `ctx` is about to be freed anyway.

aimoff replied:
> these functions are safe to pass NULL to

I'm not sure but it depends on current libaribcaption's implementation (not on the spec).
I woule like to keep the checks.

I'll accept latter suggestion.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1166 @@
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    aribcaption_flush(avctx);
```
It shouldn't be necessary to call `flush` within `close`.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1166 @@
+    ARIBCaptionContext *ctx = userdata;
+    int lvl;
+
+    if (ctx->decoder != NULL) {
```
Any particular reason for this check?

aimoff replied:
I don't remenber the details, but I caught some crashes without the check.
Probably, the callback might be called after the context has been freed.

rcombs reviewed (commented):
A few more comments, mostly regarding docs.

aimoff committed:
reflect review results

Documentation is still in progress

aimoff committed:
change bitmap positioning of profile C

The document is slightly changed.

aimoff committed:
fix misread comment for flush function

aimoff committed:
minor change

Don't calculate position of single rect for profile C

aimoff commented:
I believe almost done.
I plan to submit the patch set to ML tomorrow.

rcombs commented on libavcodec/libaribcaption.c:
```diff
@@ -0,0 +1,1171 @@
+        ctx->font_size = 16;
+        break;
+    default:
+        av_log(avctx, AV_LOG_ERROR, "Unknown or unsupported profile set.\n");
+        return AVERROR(EINVAL);
```
Oh, you'll want to provide an AVOption to set a default profile, like in the existing decoder; default it to `a` since that's what most streams are.

aimoff replied:
Does it mean to apply simiar patch as following?
https://lists.ffmpeg.org/pipermail/ffmpeg-devel/2022-November/303578.html

I don't undestand your intention.
On current lavf/mpegts.c code, such streams are not treated as ARIB caption.
If the profile is unkonwn, the codec id is not set, and libarib* decoder will not be called.
https://github.com/rcombs/FFmpeg/blob/5bad4856035ca5ed571e9d7d9b1d503a5c9ef0a5/libavformat/mpegts.c#L2138-L2181

rcombs replied:
It's possible to get arib caption streams from other containers or sources (eg matroska now), or to manually instruct ffmpeg to treat a stream as containing a particular codec if it isn't detected properly (eg due to missing descriptors).

aimoff replied:
I don't prefer such way.
Of course you can submit another patch after this patch set will be merged.

aimoff replied:
I don't know actual code and/or status of matroska.
But I prefer such workaround will be treated in libavformat.
I think more flexible decision can be made in libavformat.
